### ISSUE TITLE
Question about prefix embedding.

### ISSUE THREAD

**Issue by jiaruHithub** [minicheshire/Robust-Prefix-Tuning#3]
Hi, thanks for your novel work and release the code!
Here I have a few question about prefix embedding. I'm forward your anwser!
the batch of text classification is orgnaized as :
```
[[text1],
[text2],
[text3],
...
[text_n]]
```
where ```n``` is batch_size
when we conduct prefix-tuning, we get prefix weight, they concat with text embedding
```
[  [prefix1]    |    [text1],
   [prefix2]    |    [text2],
   [prefix3]    |    [text3], 
...
   [prefix_n]    |    [text_n]]
```
Is this cancat way right or same as the paper?
I want to know ```prefix1``` to ```prefix_n``` is same or different？
If they are same, in the backward propagarion, all prefix_n will be gradient descent in same value?
Thanks for your anwser!

**Comment by minicheshire** [comment]
Hi! I think you are interpreting the prefixes (prefix1 ~ prefixn) as instance-dependent "soft tokens", which is not exactly what we are doing in the paper.  My clarification is in the following:

1. <Key, Value> pairs at each layer, instead of only "soft tokens" at the input

In our paper, we leverage the prefix-tuning method in "[Prefix-Tuning: Optimizing Continuous Prompts for Generation](https://arxiv.org/abs/2101.00190)". So the "prefixes" (both the original and the robust ones in our paper) are in fact the <Key, Value> pairs at the starting positions (P\_idx defined in the Prefix-Tuning paper) of each layer, instead of soft tokens at the input of the bottom layer. 

Intuitively, the <Key, Value> pairs at the starting positions steer the LM towards desired behavior as they are attended to by all the subsequent tokens. Another formal expression of prefix-tuning can be found in Eq. 5 and Eq.7 in "[Towards a Unified View of Parameter-Efficient Transfer Learning](https://arxiv.org/pdf/2110.04366.pdf)".
 
2. Shared, Instead of "instance-dependent"

The original prefix (<Key, Value> pairs parameterized by $\theta$) follows "[Prefix-Tuning: Optimizing Continuous Prompts for Generation](https://arxiv.org/abs/2101.00190)", which is the same for all text inputs (during both the training and the testing phases). 

The robust prefixes (<Key, Value> pairs parameterized by $\Psi$) proposed in our paper are separately optimized for each batch of test inputs at inference time. So our robust prefixes are different across batches but the same for all inputs in a test batch.

Hope this helps. I'm also willing to explain more if you still have questions :)
